init stage 2/3 implementation

fuzzard opened the pull request:
## Description
Implement stagetwo and stagethree platform init.
Relocate tvos/macos service init to platform initstagetwo as required.

Application startup system logging has a platform override.

## Motivation and Context
Fix existing issues with components for tvos/osx builds. Provide ability for other platforms to do platform specific init at later stages than stage one

Stagetwo is done after settings have been init, so any non gui/windowing related services could/should be done here
Stagethree is done after window init. Any services/components that depend on gui/windowing can be done here. 

Should allow an easier definition of services for potential headless operation.

As part of this, streamline some other functions in Application, and move dvdmediadetect into Servicemanager

## How Has This Been Tested?
Runtime test OSX

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

Commit:
implement Platform specific InitStageTwo and Three

This allows us to add platform specific init functionality at two stages.
StageTwo allows init of services/components that do not depend/use GUI/Windowing
but require other components like settings

StageThree allows late init of services/components that require GUI/Windowing
If headless comes to fruition, any platform specific service that does not suit
can be placed here

Commit:
[Application] remove preflight

Roll preflight into platform initstageone

Commit:
[Application] relocate dvdmediadetect to servicemanager

remove 2 under utilised methods (start/stopservices). relocate the
start of dvdmediadetect to initstagetwo (service init)

sy6sy2 reviewed (approved):
Runtime tested on tvOS. Fix the issue pointed by https://github.com/xbmc/xbmc/pull/19460.

If merged, I guess we can close https://github.com/xbmc/xbmc/pull/19460 and https://github.com/xbmc/xbmc/pull/19522.

lrusak commented on xbmc/ServiceManager.cpp:
```diff
@@ -148,6 +151,16 @@ bool CServiceManager::InitStageTwo(const CAppParamParser &params, const std::str
+#if !defined(TARGET_WINDOWS) && defined(HAS_DVD_DRIVE)
+  // Start Thread for DVD Mediatype detection
+  CLog::Log(LOGINFO, "start dvd mediatype detection");
+  m_DetectDVDType.reset(new MEDIA_DETECT::CDetectDVDMedia());
```
```suggestion
  m_DetectDVDType = std::make_unique<MEDIA_DETECT::CDetectDVDMedia>();
```

garbear replied:
is the reason to hide `new` inside the unique ptr?

lrusak commented on xbmc/ServiceManager.cpp:
```diff
@@ -148,6 +151,16 @@ bool CServiceManager::InitStageTwo(const CAppParamParser &params, const std::str
 
+#if !defined(TARGET_WINDOWS) && defined(HAS_DVD_DRIVE)
+  // Start Thread for DVD Mediatype detection
+  CLog::Log(LOGINFO, "start dvd mediatype detection");
```
Can you improve this log message? (I know it's just copied from where it was)

lrusak commented on xbmc/ServiceManager.cpp:
```diff
@@ -190,6 +206,12 @@ void CServiceManager::DeinitStageTwo()
   init_level = 1;
 
+#if !defined(TARGET_WINDOWS) && defined(HAS_DVD_DRIVE)
+  CLog::Log(LOGINFO, "stop dvd detect media");
```
same with this log message

lrusak commented on xbmc/platform/Platform.h:
```diff
@@ -23,17 +23,33 @@ class CPlatform
-  virtual bool Init();
+  virtual bool Init() { return true; };
+
+  /**\brief Called at a middle stage of application startup
```
Can you note that logging will be available here (it is available in stage one but may not actually log anything because `CLog::Initialize` isn't called till after stageone. Platforms that use special loggers (android, osx) may show the logging from initstageone but other platforms that only use the log file logging won't.

lrusak commented on xbmc/platform/Platform.h:
```diff
@@ -23,17 +23,33 @@ class CPlatform
    * or initialisation (like setting environment variables for example)
    */
-  virtual bool Init();
+  virtual bool Init() { return true; };
```
might be worth it to update this to `InitStageOne`?

sy6sy2 replied:
I thought the same thing when I saw this line.  I vote for it.

fuzzard replied:
I contemplated it, then I said stuff it. Haha. I'll go back and do it though

lrusak reviewed (commented):
Generally looks ok. Greate work :100: 

I've left a few small comments. I don't like the ifdefs introduced into servicemanager but we should be able to add a instance wrapper for that at a later time.

garbear commented on xbmc/ServiceManager.h:
```diff
@@ -168,4 +177,7 @@ class CServiceManager
   std::unique_ptr<CDatabaseManager> m_databaseManager;
   std::unique_ptr<CMediaManager> m_mediaManager;
+#if !defined(TARGET_WINDOWS) && defined(HAS_DVD_DRIVE)
+  std::unique_ptr<MEDIA_DETECT::CDetectDVDMedia> m_DetectDVDType;
```
From one dumping ground to another. It would be a good time to consider a new home for anything here, but not required for the PR.

fuzzard replied:
From what I've gathered, servicebroker was just a stepping stone to greater things which just haven't materialised. I still think it helps to get everything here to give a clear picture of what to do with the DI stuff.
I've tried comprehending jimfcarrols DI stuff, but it hasn't sunk in how I could do it as yet. Maybe one day, haha

garbear replied:
@jimfcarroll's DI stuff improves our tangle of dependencies by replacing imperative relationships with declarative relationships. What we truly need is a decoupling of dependencies. However, it's a shame that "how it all should be" got in the way of the tangible benefits of DI for the parts that don't need decoupling.

garbear reviewed (approved):
This PR contains some great refactoring! I like the documentation you added for the 2nd and 3rd stage.

Commit:
rename init to initstageone

Commit:
[Application] remove platform specific logging to Platform

System startup logging is extracted out into its own method, and platform specific
logging done here has been pulled into a platform override

fuzzard commented:
Ok, couple changes, fixed @lrusak 's comments. Init renamed to initstageone for consistency

Ive also pulled out the system startup log section from CApplication::Create. As part of that, ive provided a platform override for individual platforms to output their unique device info. Just removes ifdefs in application.cpp

Im VERY open to function naming opinions here. let me know if something else is better.

lrusak commented on xbmc/platform/android/PlatformAndroid.cpp:
```diff
@@ -34,3 +37,17 @@ bool CPlatformAndroid::Init()
+  bool extready = CXBMCApp::GetExternalStorage(extstorage);
+  CLog::Log(LOGINFO, "External storage path = %s; status = %s", extstorage.c_str(),
+            extready ? "ok" : "nok");
+}
```
new line at end of file

lrusak commented on xbmc/platform/win32/WIN32Util.cpp:
```diff
@@ -1521,3 +1521,20 @@ HDR_STATUS CWIN32Util::GetWindowsHDRStatus()
+  else
+    CLog::Log(LOGINFO, "Display HDR capable is detected and Windows HDR switch is %s",
+              (hdrStatus == HDR_STATUS::HDR_ON) ? "ON" : "OFF");
+}
```
new line at end of file

lrusak reviewed (commented):
A couple more comments.

One thing I will say is that when getting others to review and they leave a review it's typically good to add extra commits (either using `--fixup` or not) after the reviewed section. That way when you request a review again the reviewer can see exactly what changed from your first review rather than having to dig through the entire PR again. Then when all the comments are addressed you can squash.

I would also be careful when increasing the scope like you have done. Sometimes it can delay the review even more. I know it doesn't seem like much but when already touching large parts of code it may be best to leave for a follow up PR. Just some food for thought :smiley:

fuzzard commented:
yeah, the scope thing was a concern. I can split it out into another PR. doesnt really bother me too much if thats the preference.

I had them as fixups, but just rebased. i'll keep that in mind for the future

fuzzard commented:
Jenkins build this please

howie-f commented on xbmc/platform/win32/PlatformWin32.cpp:
```diff
@@ -35,3 +35,8 @@ bool CPlatformWin32::Init()
   return true;
 }
+
+void CPlatformWin10::PlatformSyslog()
```
`CPlatformWin32` ?

fuzzard commented:
jenkins build this please

fuzzard commented:
Jenkins build this please

fuzzard commented:
any further comments, or good to go?

lrusak commented on xbmc/platform/android/PlatformAndroid.cpp:
```diff
@@ -34,3 +37,17 @@ bool CPlatformAndroid::Init()
+{
+  CLog::Log(
+      LOGINFO,
+      "Product: %s, Device: %s, Board: %s - Manufacturer: %s, Brand: %s, Model: %s, Hardware: %s",
```
please update to current log standards using `{}`

lrusak commented on xbmc/platform/win32/WIN32Util.cpp:
```diff
@@ -1521,3 +1521,20 @@ HDR_STATUS CWIN32Util::GetWindowsHDRStatus()
+  CLog::Log(LOGINFO, "Running with %s rights",
+            (IsCurrentUserLocalAdministrator() == TRUE) ? "administrator"
+                                                        : "restricted");
+  CLog::Log(LOGINFO, "Aero is %s", (g_sysinfo.IsAeroDisabled() == true) ? "disabled" : "enabled");
```
please update to current log standards using `{}`

lrusak commented on xbmc/Application.cpp:
```diff
@@ -5052,6 +4910,80 @@ void CApplication::SetLoggingIn(bool switchingProfiles)
+  CLog::Log(LOGINFO, "Using %s %s x%d", buildType.c_str(), CSysInfo::GetAppName().c_str(),
+            g_sysinfo.GetXbmcBitness());
+  CLog::Log(
+      LOGINFO, "%s compiled %s by %s for %s %s %d-bit %s (%s)", CSysInfo::GetAppName().c_str(),
```
please update to current log standards using `{}`

lrusak commented on xbmc/Application.cpp:
```diff
@@ -5052,6 +4910,80 @@ void CApplication::SetLoggingIn(bool switchingProfiles)
+  if (!g_sysinfo.GetManufacturerName().empty())
+    deviceModel = g_sysinfo.GetManufacturerName() + " " + (deviceModel.empty() ? std::string("device") : deviceModel);
+  if (!deviceModel.empty())
+    CLog::Log(LOGINFO, "Running on %s with %s, kernel: %s %s %d-bit version %s",
```
please update to current log standards using `{}`

lrusak commented on xbmc/Application.cpp:
```diff
@@ -5052,6 +4910,80 @@ void CApplication::SetLoggingIn(bool switchingProfiles)
+              g_sysinfo.GetKernelName().c_str(), g_sysinfo.GetKernelCpuFamily().c_str(),
+              g_sysinfo.GetKernelBitness(), g_sysinfo.GetKernelVersionFull().c_str());
+  else
+    CLog::Log(LOGINFO, "Running on %s, kernel: %s %s %d-bit version %s",
```
please update to current log standards using `{}`

lrusak commented on xbmc/Application.cpp:
```diff
@@ -5052,6 +4910,80 @@ void CApplication::SetLoggingIn(bool switchingProfiles)
+              g_sysinfo.GetKernelCpuFamily().c_str(), g_sysinfo.GetKernelBitness(),
+              g_sysinfo.GetKernelVersionFull().c_str());
+
+  CLog::Log(LOGINFO, "FFmpeg version/source: %s", av_version_info());
```
please update to current log standards using `{}`

lrusak commented on xbmc/Application.cpp:
```diff
@@ -5052,6 +4910,80 @@ void CApplication::SetLoggingIn(bool switchingProfiles)
+
+  std::string cpuModel(CServiceBroker::GetCPUInfo()->GetCPUModel());
+  if (!cpuModel.empty())
+    CLog::Log(LOGINFO, "Host CPU: %s, %d core%s available", cpuModel.c_str(),
```
please update to current log standards using `{}`

lrusak commented on xbmc/Application.cpp:
```diff
@@ -5052,6 +4910,80 @@ void CApplication::SetLoggingIn(bool switchingProfiles)
+              CServiceBroker::GetCPUInfo()->GetCPUCount(),
+              (CServiceBroker::GetCPUInfo()->GetCPUCount() == 1) ? "" : "s");
+  else
+    CLog::Log(LOGINFO, "%d CPU core%s available", CServiceBroker::GetCPUInfo()->GetCPUCount(),
```
please update to current log standards using `{}`

lrusak commented on xbmc/Application.cpp:
```diff
@@ -5052,6 +4910,80 @@ void CApplication::SetLoggingIn(bool switchingProfiles)
+  CSpecialProtocol::LogPaths();
+
+  std::string executable = CUtil::ResolveExecutablePath();
+  CLog::Log(LOGINFO, "The executable running is: %s", executable.c_str());
```
please update to current log standards using `{}`

lrusak commented on xbmc/Application.cpp:
```diff
@@ -5052,6 +4910,80 @@ void CApplication::SetLoggingIn(bool switchingProfiles)
+  CLog::Log(LOGINFO, "The executable running is: %s", executable.c_str());
+  std::string hostname("[unknown]");
+  m_ServiceManager->GetNetwork().GetHostName(hostname);
+  CLog::Log(LOGINFO, "Local hostname: %s", hostname.c_str());
```
please update to current log standards using `{}`

lrusak commented on xbmc/Application.cpp:
```diff
@@ -5052,6 +4910,80 @@ void CApplication::SetLoggingIn(bool switchingProfiles)
+  CLog::Log(LOGINFO, "Local hostname: %s", hostname.c_str());
+  std::string lowerAppName = CCompileInfo::GetAppName();
+  StringUtils::ToLower(lowerAppName);
+  CLog::Log(LOGINFO, "Log File is located: %s.log",
```
please update to current log standards using `{}`

lrusak reviewed (changes requested):
I hate to do it but can you pretty please update the logging to current standards. This also involves dropping the `.c_str()` where applicable. Thanks!

lrusak commented:
jenkins build this please

fuzzard commented:
Jenkins build this please

lrusak commented:
jenkins build this please

fuzzard commented:
jenkins build this please

fuzzard commented:
This good to go now? Had some troubles getting Jenkins to greenlight everything, but all good now from that perspective

howie-f commented on xbmc/Application.cpp:
```diff
@@ -4999,6 +4857,80 @@ void CApplication::SetLoggingIn(bool switchingProfiles)
+  m_ServiceManager->GetPlatform().PlatformSyslog();
+
+#if defined(__arm__) || defined(__aarch64__)
+  if (CServiceBroker::GetCPUInfo()->GetCPUFeatures() & CPU_FEATURE_NEON)
+    CLog::Log(LOGINFO, "ARM Features: Neon enabled");
+  else
+    CLog::Log(LOGINFO, "ARM Features: Neon disabled");
```
```
CLog::Log(LOGINFO, "ARM Features: Neon {}",
  (CServiceBroker::GetCPUInfo()->GetCPUFeatures() & CPU_FEATURE_NEON) ? "enabled" : "disabled");
```

howie-f commented on xbmc/platform/darwin/osx/PlatformDarwinOSX.cpp:
```diff
@@ -8,27 +8,42 @@
 
   CCocoaPowerSyscall::Register();
 
+  std::string install_path;
+
+  install_path = CUtil::GetHomePath();
```
`std::string install_path(CUtil::GetHomePath());`

howie-f reviewed (approved):
no negative impact seen at linux runtime
